mlir/test/mlir-miopen-driver/e2e/conv2d_harness_cpu_verifier_kyxc_nhwc_nhwk_f16.mlir

```diff
@@ -0,0 +1,274 @@
+    %0 = memref_cast %arg0 : memref<128x3x3x8xf32> to memref<*xf32>
+    %1 = memref_cast %arg1 : memref<128x32x32x8xf32> to memref<*xf32>
+    %2 = memref_cast %arg2 : memref<128x30x30x128xf32> to memref<*xf32>
+    %c1_i32 = constant 1 : i32
+    %c1_i32_0 = constant 1 : i32
+    %c0_i32 = constant 0 : i32
+    %c0_i32_1 = constant 0 : i32
+    %c1_i32_2 = constant 1 : i32
+    %c1_i32_3 = constant 1 : i32
+
+    // set up constant indices
+    %c0 = constant 0 : index
+    %c1 = constant 1 : index
+    %c2 = constant 2 : index
+    %c3 = constant 3 : index
+
+    // set up constants (ascii code) for layout letters
+    %c107_i8 = constant 107 : i8
+    %c99_i8 = constant 99 : i8
+    %c121_i8 = constant 121 : i8
+    %c120_i8 = constant 120 : i8
+    %c110_i8 = constant 110 : i8
+    %c104_i8 = constant 104 : i8
+    %c119_i8 = constant 119 : i8
+
+    // allocate memory for layouts
+    %3 = alloca() : memref<4xi8>
+    %4 = alloca() : memref<4xi8>
+    %5 = alloca() : memref<4xi8>
+
+    // store layouts
+    store %c107_i8, %3[%c0] : memref<4xi8>
+    store %c121_i8, %3[%c1] : memref<4xi8>
+    store %c120_i8, %3[%c2] : memref<4xi8>
+    store %c99_i8, %3[%c3] : memref<4xi8>
+    store %c110_i8, %4[%c0] : memref<4xi8>
+    store %c104_i8, %4[%c1] : memref<4xi8>
+    store %c119_i8, %4[%c2] : memref<4xi8>
+    store %c99_i8, %4[%c3] : memref<4xi8>
+    store %c110_i8, %5[%c0] : memref<4xi8>
+    store %c104_i8, %5[%c1] : memref<4xi8>
+    store %c119_i8, %5[%c2] : memref<4xi8>
+    store %c107_i8, %5[%c3] : memref<4xi8>
+    %6 = memref_cast %3 : memref<4xi8> to memref<*xi8>
+    %7 = memref_cast %4 : memref<4xi8> to memref<*xi8>
+    %8 = memref_cast %5 : memref<4xi8> to memref<*xi8>
+    call @mcpuConv2d(%0, %1, %2, %6, %7, %8, %c1_i32, %c1_i32_0, %c0_i32, %c0_i32_1, %c1_i32_2, %c1_i32_3) : (memref<*xf32>, memref<*xf32>, memref<*xf32>, memref<*xi8>, memref<*xi8>, memref<*xi8>, i32, i32, i32, i32, i32, i32) -> ()
+    return
+  }
+
+  func private @mcpuConv2d(memref<*xf32>, memref<*xf32>, memref<*xf32>, memref<*xi8>, memref<*xi8>, memref<*xi8>, i32, i32, i32, i32, i32, i32)
+
+  func @verify_results(%arg0: memref<128x30x30x128xf32>, %arg1: memref<128x30x30x128xf32>) {
+    %c0 = constant 0 : index
+    %0 = alloca() : memref<1xi32>
+    %c0_i32 = constant 0 : i32
+    %c1_i32 = constant 1 : i32
+    store %c1_i32, %0[%c0] : memref<1xi32>
+    %c1 = constant 1 : index
+    %c128 = constant 128 : index
+    %c30 = constant 30 : index
+    %c30_0 = constant 30 : index
+    %c128_1 = constant 128 : index
+    scf.for %arg2 = %c0 to %c128 step %c1 {
+      scf.for %arg3 = %c0 to %c30 step %c1 {
+        scf.for %arg4 = %c0 to %c30_0 step %c1 {
+          scf.for %arg5 = %c0 to %c128_1 step %c1 {
+            %2 = load %arg0[%arg2, %arg3, %arg4, %arg5] : memref<128x30x30x128xf32>
+            %3 = load %arg1[%arg2, %arg3, %arg4, %arg5] : memref<128x30x30x128xf32>
+            %4 = cmpf une, %2, %3 : f32
```

Comment:
What is the precision of this comparison?

Reply:
With fp16, an Integer less than -2048 or greater than 2048 round to an even number. So the comparison error can be at least 1. In practice, I haven't seen this happened.

Reply:
> So the comparison error can be at least 1. In practice, I haven't seen this happened.

Does this 1 comparison margin too large to judge the correctness? In MIOpen, the comparison margin must be less than 1e-5, which is way smaller than 1.

Reply:
I think MIOpen uses floating point values. I borrowed the idea from the modular_convolution repo where random integers in [-5,5] are used, hoping that out-of-bound cases wouldn't happen. Chao Liu confirmed that using integers would help find more errors.